src/Symfony/Component/Messenger/Bridge/GooglePubSub/Transport/GooglePubSubReceiver.php

```diff
@@ -0,0 +1,146 @@
+use Symfony\Component\Messenger\Transport\Serialization\PhpSerializer;
+use Symfony\Component\Messenger\Transport\Serialization\SerializerInterface;
+
+/**
+ * @author Richard Henkenjohann <richardhenkenjohann@googlemail.com>
+ */
+class GooglePubSubReceiver implements ReceiverInterface
+{
+    private const MESSAGE_ATTRIBUTE_NAME = 'X-Symfony-Messenger';
+
+    private $subscription;
+    private $serializer;
+
+    /** @var array[] */
+    private $buffer = [];
+
+    public function __construct(Subscription $subscription, SerializerInterface $serializer = null)
+    {
+        $this->subscription = $subscription;
+        $this->serializer   = $serializer ?? new PhpSerializer();
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function get(): iterable
+    {
+        $gpsEnvelope = $this->getMessage();
+        if (null === $gpsEnvelope) {
+            return;
+        }
+
+        try {
+            $envelope = $this->serializer->decode([
+                'body' => $gpsEnvelope['body'],
+                'headers' => $gpsEnvelope['headers'],
+            ]);
+        } catch (MessageDecodingFailedException $exception) {
+            $this->subscription->delete($gpsEnvelope['id']);
+
+            throw $exception;
+        }
+
+        yield $envelope->with(new GooglePubSubReceivedStamp($gpsEnvelope['id'], $gpsEnvelope['ackId']));
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function ack(Envelope $envelope): void
+    {
+        $stamp = $this->findReceivedStamp($envelope);
+
+        $this->subscription->acknowledge(new Message([], ['ackId' => $stamp->getAckId()]));
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function reject(Envelope $envelope): void
+    {
+        $this->ack($envelope);
+    }
+
+    private function findReceivedStamp(Envelope $envelope): GooglePubSubReceivedStamp
+    {
+        /** @var GooglePubSubReceivedStamp|null $stamp */
+        $stamp = $envelope->last(GooglePubSubReceivedStamp::class);
+
+        if (null === $stamp) {
+            throw new LogicException('No GooglePubSubReceivedStamp found on the Envelope.');
+        }
+
+        return $stamp;
+    }
+
+    private function getMessage(): ?array
+    {
+        foreach ($this->getNextMessages() as $message) {
+            return $message;
+        }
+
+        return null;
+    }
+
+    private function getNextMessages(): \Generator
+    {
+        yield from $this->getPendingMessages();
+        yield from $this->getNewMessages();
+    }
+
+    private function getPendingMessages(): \Generator
+    {
+        while (!empty($this->buffer)) {
+            yield array_shift($this->buffer);
+        }
+    }
+
+    private function getNewMessages(): \Generator
+    {
+        $this->fetchMessages();
+
+        yield from $this->getPendingMessages();
+    }
+
+    private function fetchMessages(): void
+    {
+        foreach ($this->subscription->pull() as $message) {
```

Comment:
By default, the SDK returns 1000 messages.
It sound like a lot to store in memory and to ack before the deadline.
I would make it parameterizable.